src/backend/binaryen/utils.ts

```diff
@@ -2016,4 +2028,312 @@ export namespace FunctionalFuncs {
+        vars.push(binaryen.i32);
+        const loopIndexValue = module.local.get(startIdx, binaryen.i32);
+        const codesArray = binaryenCAPI._BinaryenStructGet(
+            module.ptr,
+            0,
+            arrayBufferRef,
+            arrayBufferTypeInfo.typeRef,
+            false,
+        );
+        /* Put elem in arraybuffer */
+        const loopLabel = 'for_label';
+        const loopCond = module.i32.lt_s(loopIndexValue, lengthRef);
+        const loopIncrementor = module.local.set(
+            startIdx,
+            module.i32.add(loopIndexValue, module.i32.const(1)),
+        );
+        const loopBody: binaryen.ExpressionRef[] = [];
+        loopBody.push(
+            binaryenCAPI._BinaryenArraySet(
+                module.ptr,
+                codesArray,
+                loopIndexValue,
+                module.i32.load8_s(
+                    0,
+                    1,
+                    module.i32.add(offsetValueRef, loopIndexValue),
+                ),
+            ),
+        );
+        const flattenLoop: FlattenLoop = {
+            label: loopLabel,
+            condition: loopCond,
+            statements: module.block(null, loopBody),
+            incrementor: loopIncrementor,
+        };
+        stmts.push(
+            module.loop(
+                loopLabel,
+                FunctionalFuncs.flattenLoopStatement(
+                    module,
+                    flattenLoop,
+                    SemanticsKind.FOR,
+                ),
+            ),
+        );
+        return module.block(null, stmts);
+    }
+
+    export function parseNativeSignatureConversion(
+        fromTypes: ValueType[],
+        toTypes: ValueType[],
+    ) {
+        /* fromTypes is the wrapper functions' parameter types, toTypes is the real functions's parameter types */
+        if (fromTypes.length !== toTypes.length) {
+            throw new Error(
+                `NativeSignature's parameter length must match real function's parameter length`,
+            );
+        }
+        const convertRules: NativeSignatureConversion[] = [];
+        for (let i = 0; i < fromTypes.length; i++) {
+            switch (fromTypes[i].kind) {
+                case ValueTypeKind.OBJECT: {
+                    const className = (fromTypes[i] as ObjectType).meta.name;
+                    switch (className) {
+                        case BuiltinNames.ARRAYBUFFER: {
+                            switch (toTypes[i].kind) {
+                                case ValueTypeKind.INT: {
+                                    convertRules.push(
+                                        NativeSignatureConversion.ARRAYBUFFER_TO_I32,
+                                    );
+                                    break;
+                                }
+                                default: {
+                                    convertRules.push(
+                                        NativeSignatureConversion.INVALID,
```

Comment:
Too many branches, how about just process valid conversion patterns and return immediately, then all invalid pattern will fall to a final false branch?

Reply:
Great idea, I will change it.